nova_vm/src/engine/bytecode/executable/function_declaration_instantiation.rs

```diff
@@ -0,0 +1,334 @@
+    for d in function_body_var_scoped_declarations(body) {
+        // a. If d is neither a VariableDeclaration nor a ForBinding nor a BindingIdentifier, then
+        if let VarScopedDeclaration::Function(d) = d {
+            // i. Assert: d is either a FunctionDeclaration, a GeneratorDeclaration, an AsyncFunctionDeclaration, or an AsyncGeneratorDeclaration.
+            // ii. Let fn be the sole element of the BoundNames of d.
+            let f_name = d.id.as_ref().unwrap().name.clone();
+            // iii. If functionNames does not contain fn, then
+            //   1. Insert fn as the first element of functionNames.
+            //   2. NOTE: If there are multiple function declarations for the same name, the last declaration is used.
+            //   3. Insert d as the first element of functionsToInitialize.
+            functions.insert(f_name, d);
+        }
+    }
+
+    // 15. Let argumentsObjectNeeded be true.
+    // 16. If func.[[ThisMode]] is lexical, then
+    //   a. NOTE: Arrow functions never have an arguments object.
+    //   b. Set argumentsObjectNeeded to false.
+    // 17. Else if parameterNames contains "arguments", then
+    //   a. Set argumentsObjectNeeded to false.
+    // 18. Else if hasParameterExpressions is false, then
+    //   a. If functionNames contains "arguments" or lexicalNames contains "arguments", then
+    //     i. Set argumentsObjectNeeded to false.
+    let arguments_object_needed = !is_lexical
+        && !parameter_names.contains("arguments")
+        && (has_parameter_expressions
+            || (!functions.contains_key("arguments")
+                && !function_body_lexically_declared_names(body)
+                    .contains(&Atom::from("arguments"))));
+
+    // 19. If strict is true or hasParameterExpressions is false, then
+    //   a. NOTE: Only a single Environment Record is needed for the parameters, since calls to eval in strict mode code cannot create new bindings which are visible outside of the eval.
+    //   b. Let env be the LexicalEnvironment of calleeContext.
+    // 20. Else,
+    //   a. NOTE: A separate Environment Record is needed to ensure that bindings created by direct eval calls in the formal parameter list are outside the environment where parameters are declared.
+    //   b. Let calleeEnv be the LexicalEnvironment of calleeContext.
+    //   c. Let env be NewDeclarativeEnvironment(calleeEnv).
+    //   d. Assert: The VariableEnvironment of calleeContext and calleeEnv are the same Environment Record.
+    //   e. Set the LexicalEnvironment of calleeContext to env.
+    if !strict && has_parameter_expressions {
+        ctx.exe
+            .add_instruction(Instruction::EnterDeclarativeEnvironment);
+    }
+
+    // 21. For each String paramName of parameterNames, do
+    // NOTE: The behavior should not depend on the order in which the parameter
+    // names are iterated, so it's fine for `parameter_names` to be a set.
+    for param_name in &parameter_names {
```

Comment:
**thought:** At some point it might be a good idea to create a `InitializeDeclarativeEnvironment` kind of instruction that would point to a "declarative environment template" which has (mostly?) uninitialized bindings for all the names that will be needed. Performing that instruction would check that the current declarative environment is empty, and then replace its Bindings Map with a clone of a Bindings Map that comes from the Vm context (or that place where we save bindings etc. anyhow).

Essentially, a quicker way to prep the declarative environment since we know beforehand which names we're going to define in it.